### PR TITLE
adam optimizer support lazy mode

### PR DESCRIPTION
**Pull request by jacquesqiao**
in my test case, in distribute training, when the parameter size is 800M, the adam optimizer will be 10 more time faster.

### CODE REVIEW

**Review by typhoonzero — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 1]
```diff
@@ -351,11 +357,27 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
-      for_range(functor);
+          grad_merge.rows().size(), lazy_mode);
+      VLOG(3) << "lazy_mode :" << lazy_mode;
+      if (lazy_mode) {
+        std::vector<int64_t> id_vector;
+        size_t row_count = grad_merge.rows().size();
+        std::vector<int64_t> cpu_rows(grad_merge.rows());
+        for (size_t row_index = 0; row_index < row_count; ++row_index) {
+          for (size_t offset = 0; offset < row_numel; ++offset) {
+            size_t i = cpu_rows[row_index] * row_numel + offset;
```
Comment (chengduoZH) [inline paddle/fluid/operators/optimizers/adam_op.h:371]: What do you want to with `cpu_rows`? It is defined but not initialized.

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 2]
```diff
@@ -351,11 +357,27 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
+            id_vector.push_back(i);
+          }
+        }
+        platform::ForRangeIn<DeviceContext> for_range_in(
+            static_cast<const DeviceContext&>(ctx.device_context()), id_vector);
```
Comment (chengduoZH) [inline paddle/fluid/operators/optimizers/adam_op.h]: Note that the `for_range_in ` is in stack memory, once the function return, the memory will be released.
You define `range_` [here](https://github.com/PaddlePaddle/Paddle/pull/14889/files#diff-f3be7a1b3192291a4c06a311d5e643abR135), and it is used by CUDA Kernel, if the CUDA Kernel doesn't start to execute before the current function returns, the `range_` will be released, it is dangerous.

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 3]
```diff
@@ -351,11 +357,27 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
+        for_range_in(functor);
```
Comment (chengduoZH) [inline paddle/fluid/operators/optimizers/adam_op.h]: You can use `platform::ForRange` do above task.

File: paddle/fluid/operators/optimizers/adam_op.h (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/optimizers/adam_op.h:371]: cpu_rows is inited https://github.com/PaddlePaddle/Paddle/pull/14889/files#diff-e616ec13ecb3784788f198a7812b4a0aR365
```cpp
std::vector<int64_t> cpu_rows(grad_merge.rows());
```

File: paddle/fluid/operators/optimizers/adam_op.h (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/optimizers/adam_op.h]: No, ForRange can not to this task.

File: paddle/fluid/operators/optimizers/adam_op.h (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/operators/optimizers/adam_op.h]: This is a problem

File: python/paddle/fluid/optimizer.py [hunk 4]
```diff
@@ -641,9 +641,14 @@ class AdamOptimizer(Optimizer):
...
-        regularization: A Regularizer, such as
-                        fluid.regularizer.L2DecayRegularizer.
+        regularization: A Regularizer, such as fluid.regularizer.L2DecayRegularizer.
         name: A optional name prefix.
+        lazy_mode(bool: false): The official Adam algorithm has two moving-average accumulators
+        the accumulators are updated at every step. Every element of the two moving-average is updated
+        in both dense mode and sparse mode. If the size of parameter is very large, then the update
+        may be very slow. The lazy mode only update the element that has gradient is the current
+        mini-batch, so it will be much more faster. But this mode has different semantics with the
+        original Adam algorithm and may lead to different result.
```
Comment (chengduoZH) [inline python/paddle/fluid/optimizer.py:651]: Do you have the performance report in lazy_mode? If so, you can post the data in this PR.

Comment (jacquesqiao) [reply python/paddle/fluid/optimizer.py:651]: Done

**Review by chengduoZH — APPROVED** [review]

File: python/paddle/fluid/optimizer.py [hunk 5]
```diff
@@ -641,9 +641,14 @@ class AdamOptimizer(Optimizer):
...
         beta2 (float): The exponential decay rate for the 2nd moment estimates.
         epsilon (float): a small float value for numerical stability.
-        regularization: A Regularizer, such as
-                        fluid.regularizer.L2DecayRegularizer.
+        regularization: A Regularizer, such as fluid.regularizer.L2DecayRegularizer.
         name: A optional name prefix.
+        lazy_mode(bool: false): The official Adam algorithm has two moving-average accumulators
+        the accumulators are updated at every step. Every element of the two moving-average is updated
+        in both dense mode and sparse mode. If the size of parameter is very large, then the update
+        may be very slow. The lazy mode only update the element that has gradient is the current
```
Comment (qingqing01) [inline python/paddle/fluid/optimizer.py:649]: The lazy mode only update -> The lazy mode only updates
is the current -> in current

**Review by panyx0718 — APPROVED** [review]
LG API

**Review by typhoonzero — APPROVED** [review]

**Pull request merged**